apps/prairielearn/src/ee/auth/lti13/lti13Auth.ts

```diff
@@ -0,0 +1,306 @@
+    if (!lti13_instance.uid_attribute) {
+      throw error.make(500, 'LTI 1.3 instance configuration missing required UID attribute');
+    } else {
+      // Uses lodash.get to expand path representation in text to the object, like 'a[0].b.c'
+      // Reasonable default is "email"
+      // Points back to OIDC Standard Claims https://openid.net/specs/openid-connect-core-1_0.html#StandardClaims
+      uid = _get(lti13_claims, lti13_instance.uid_attribute);
+      if (!uid) {
+        // Canvas Student View does not include a uid but has a deterministic role, nicer error message
+        if (
+          lti13_claims['https://purl.imsglobal.org/spec/lti/claim/roles']?.includes(
+            'http://purl.imsglobal.org/vocab/lti/system/person#TestUser',
+          )
+        ) {
+          throw error.make(
+            403,
+            `Student View / Test user not supported. Use access modes within PrairieLearn to view as a student.`,
+          );
+        } else {
+          // Error about missing UID
+          throw error.make(
+            500,
+            `Missing UID data from LTI 1.3 login (claim ${lti13_instance.uid_attribute} missing or empty)`,
+          );
+        }
+      }
+    }
+
+    // UIN checking, if attribute defined value must be present
+    uin = null;
+    if (lti13_instance.uin_attribute) {
+      // Uses lodash.get to expand path representation in text to the object, like 'a[0].b.c'
+      // Might look like ["https://purl.imsglobal.org/spec/lti/claim/custom"]["uin"]
+      uin = _get(lti13_claims, lti13_instance.uin_attribute);
+      if (!uin) {
+        throw error.make(
+          500,
+          `Missing UIN data from LTI 1.3 login (claim ${lti13_instance.uin_attribute} missing or empty)`,
+        );
+      }
+    }
+
+    // Name checking, not an error
+    // LTI 1.3 spec defines sharing name as a MAY https://www.imsglobal.org/spec/lti/v1p3#users-and-roles
+    // but discourages (MUST NOT) using other attributes for unique identifier
+    name = null;
+    if (lti13_instance.name_attribute) {
+      // Uses lodash.get to expand path representation in text to the object, like 'a[0].b.c'
+      // Reasonable default is "name"
+      // Points back to OIDC Standard Claims https://openid.net/specs/openid-connect-core-1_0.html#StandardClaims
+      name = _get(lti13_claims, lti13_instance.name_attribute);
+    }
+
+    const userInfo = {
+      uid,
+      uin,
+      name,
+      provider: 'LTI 1.3',
+      institution_id: lti13_instance.institution_id,
+    };
+
+    // AUTHENTICATE
+    await authnLib.loadUser(req, res, userInfo);
+
+    // Record the LTI 1.3 user's subject id
+    await queryAsync(sql.update_lti13_users, {
+      user_id: res.locals.authn_user.user_id,
+      lti13_instance_id: lti13_instance.id,
+      sub: lti13_claims.sub,
+    });
+
+    // Put the LTI 1.3 claims in the session
+    req.session.lti13_claims = lti13_claims;
+    req.session.authn_lti13_instance_id = lti13_instance.id;
```

Comment:
Since this PR hadn't landed yet and I need it for the follow up work, I've included two session variables from the LTI auth. 

`req.session.lti13_claims` holds all of the LTI 1.3 info and it used for authz and course linking.
`req.session.authn_lti13_instance_id` holds the LTI 1.3 instance the authentication came from, as we don't have a way to calculate that backwards later. That feels like metadata I would expect to put into `res.locals.authn_provider = {}` but we don't have that (only `res.locals.authn_provider_name`). This is a bandaid to get the LTI stuff going in a place where the later LTI pages can reference it.

Reply:
I was hoping to save discussion of session stuff til later and would otherwise be approving this PR immediately. Some of the questions I have here:

- When/how will we expire this data? I don't want this sitting around in sessions forever.
- What happens if someone tries to initiate another linking while they're still in the middle of another linking flow? It's only a matter of time before that happens. To that end, I was imagining scoping this data by something (instance ID? something else?).

Reply:
For expiration, I propose we could put a check in a library that looked for `req.session.lti13_claims.exp` and removed the `lti13_claims` object if it's past the expired timestamp. That kind of check would be done before anything that use the LTI claims. That wouldn't help it from sitting around in sessions forever, unless there's a way to trigger that expiry somehow from the sessions?

Reply:
> I had imagined this would end up in the URL of whatever PL page we end up redirecting users to, is that not the case?

I'd prefer not to trust the referrer URL and embed some authentication data in the session. The LTI data doesn't need to be long lived (your expiration question), but there are some things we need to pass around while we're initializing or configuring the LTI linking. An example case is where someone might be authenticated via LTI through lti13_instance 1 and visits `/pl/lti13_instance/2/course_navigation`.

I'm not sure lti13 instance scoping the login data is needed. A single scope that can be overwritten by a newer LTI ingress link is an easy way to toggle. We can do the validation and error to point them to link again via the LTI if their data is invalid?

Reply:
Alternatively we can remove these from the auth PR and discuss/implement in #9098

Reply:
My preference is to remove these changes from this PR and add them again in #9098.

Reply:
Done in d3882f83c, we'll push ahead in the next PR